src/lurk/eval.rs

```diff
@@ -2086,6 +2133,112 @@ pub fn eval_letrec<F: AbstractField>() -> FuncE<F> {
     )
 }
 
+pub fn collect_mutual_env<F: AbstractField>() -> FuncE<F> {
+    func!(
+        fn collect_mutual_env(binds_tag, binds): [2] {
+            let zero = 0;
+            let invalid_form_err = EvalErr::InvalidForm;
+            match binds_tag {
+                InternalTag::Nil => {
+                    let one = 1;
+                    return (one, zero)
+                }
+                Tag::Cons => {
+                    let cons_tag = Tag::Cons;
+                    let (binding_tag, binding, binds_tag, binds) = load(binds);
+                    let (success, tail_env) = call(collect_mutual_env, binds_tag, binds);
+                    if !success {
+                        return (zero, invalid_form_err)
+                    }
+                    let binding_not_cons = sub(binding_tag, cons_tag);
+                    if binding_not_cons {
+                        return (zero, invalid_form_err)
+                    }
+                    let (var_tag, var, rest_tag, rest) = load(binding);
+                    match var_tag {
+                        Tag::Sym, Tag::Builtin, Tag::Coroutine => {
+                            let rest_tag_not_cons = sub(rest_tag, cons_tag);
+                            if rest_tag_not_cons {
+                                return (zero, invalid_form_err)
+                            }
+                            let (val_tag, val, rest_tag, _rest) = load(rest);
+                            let nil_tag = InternalTag::Nil;
+                            let rest_tag_not_nil = sub(rest_tag, nil_tag);
+                            if rest_tag_not_nil {
+                                return (zero, invalid_form_err)
+                            }
+                            let one = 1;
+                            let env = store(var_tag, var, val_tag, val, tail_env);
+                            return (one, env)
+                        }
+                    };
+                    let illegal_binding_var_err = EvalErr::IllegalBindingVar;
+                    return (zero, illegal_binding_var_err)
+                }
+            };
+            return (zero, invalid_form_err)
+        }
+    )
+}
+
+pub fn extend_env_with_mutuals<F: AbstractField>() -> FuncE<F> {
+    func!(
+        fn extend_env_with_mutuals(consumed_mutual_env, full_mutual_env, constructed_env, original_env): [1] {
```

Comment:
This function is recursive. Does it need to be tagged `partial`?

Reply:
I think this function only needs to be `partial` if we believe that a circular env can be constructed inside the VM. Since I believe that is not the case (same case as a cyclic cons list), it's fine for this to not be partial. Same case as above